Cargo.toml

```diff
@@ -21,3 +21,4 @@ handlebars = "3"
 colored = "2"
 colourado = "0.2.0"
 memchr = "2.4.1"
+ctrlc = "3.2.1"
```

Comment:
You can use [`tokio::signal::ctrl_c`](https://docs.rs/tokio/latest/tokio/signal/fn.ctrl_c.html) instead by enabling feature `signal` on dependency `tokio`.

It removes the need to use `tokio::sync::Mutex<Bool>` or `tokio::sync::Notify` altogether,

Reply:
Oh god tokio has a lot of things. Are you thinking of something like this?

```rust
tokio::spawn(async move {
    loop {
        tokio::select!(
            _ = tokio::signal::ctrl_c() => break,
            _ = async { /* scheduling loop */ } => continue,
        );
    }
});
```

Reply:
I suppose this kind of loop is totally OK, if everything in it is cancel safe.

Reply:
I think `tokio::signal::ctrl_c()` can be called multiple times and all of these `.await` on it would return on receiving one `ctrl_c` signal, however since I don't have much experience with `tokio::signal`, I am not so sure.

Reply:
The scheduling loop is probably not cancel safe. The scheduling loop is designed to check ctrl-c cancellation before fetching a new command from the queue, and I feel like awaiting on `Notify::notified` or `ctrl_c` will end up being async w.r.t the operations of the scheduling loop, making things difficult. Plus, if I'm going to refactor it as a task, then I should figure out a way to `drop(command_txs)`, etc when the loop exits.

High chance that I'm not figuring out what you imagined how to use `ctrl_c`?

Reply:
I read `Session::run` and I am pretty sure it is cancel safe.

And, you don't have to call `ctrl_c` in these `tasks`.

You can call [`tokio::task::JoinHandle:;abort`](https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#method.abort) to abort these tasks.

Reply:
Oh, for the code snippet I wrote, I was talking about the scheduling loop (the loop that calls `job_queue.next()` and sends out commands to tasks). Sorry for the confusion.

I see. I just experimented a bit and saw that calling `abort` on a task will call `drop` for all objects inside. (In hindsight this was obvious haha.) So that'll at least handle SSH sessions using its `drop` implementation, though errors will be ignored.

One thing is that, I'd like the commands that are running at the time of "ctrl-c" to run until completion. That's because even if I forcefully kill the tasks and `ssh`, the remote child process is going to continue running anyway. Then, I'd at least want to continue seeing its output. At the point "ctrl-c" has been admitted by Pegasus, neither `queue.yaml` nor `consumed.yaml` will be modified by that instance of Pegasus and I can just start up a new instance of Pegasus to run stuff again. I don't see a way to achieve this with `abort`...

Reply:
Maybe a bit off topic, but I am thinking of a different architecture here:

All `Session`s are stored in a `Arc<HashMap<String, Session>>`, with its name as the key.
Since the hosts are loaded at the startup, it is perfectly OK to store them in a `Arc<HashMap>`.

Then, upon receiving new tasks, search for the corresponding `Session` and spawn a new task to execute the job.

The new task communicates with the [`tokio::sync::Semaphore`].
When the new task is done, call [`tokio::sync::Semaphore::add_permits`].

The main task can then call [`tokio::sync::Semaphore::acquire_many`] to wait for the completion of `n` tasks and uses  [`tokio::sync::SemaphorePermit::forget`] to prevent the permit from being returned to the [`tokio::sync::Semaphore`].

By doing each job in a task instead of running every `Session` in a task, we can do graceful shutdown easily using `tokio::signal::ctrl_c`.

[`tokio::sync::Semaphore`]: https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html
[`tokio::sync::Semaphore::add_permits`]: https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html#method.add_permits
[`tokio::sync::Semaphore::acquire_many`]: https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html#method.acquire_many
[`tokio::sync::SemaphorePermit::forget`]: https://docs.rs/tokio/latest/tokio/sync/struct.SemaphorePermit.html#method.forget

Reply:
Using the architecture I described above, I believe it is also possible to use [`notify`] in `tokio::spawn` combined with [`tokio::sync::watch`]  to monitor the change for change of `queue.yaml`.

It is possible since [`tokio::sync::watch::Receiver::changed`] is cancel safe.

[`notify`]: https://docs.rs/notify/4.0.17/notify/
[`tokio::sync::watch`]: https://docs.rs/tokio/latest/tokio/sync/watch/index.html
[`tokio::sync::watch::Receiver::changed`]: https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.changed

Reply:
Upon inspection of implementation of [`tokio::signal::ctrl_c`](https://docs.rs/tokio/latest/src/tokio/signal/unix.rs.html), I found that it just uses `tokio::sync::watch::Receiver` to receive signal, so it should also be cancel safe.

Reply:
If we actually need to unload/load new hosts dynamically, then I would use `HashMap<String, Arc<Session>>` to stores the sessions.

Reply:
Very interesting suggestion! I think this will work for broadcast mode, but I'm not sure about queue mode. All in all this is a major rework of the core logic, so if the current one at least works, I think the refactoring should be done in a separate PR. Also, I feel like the suggestions about `notify` + `tokio::sync::watch` and using `tokio::signal::ctrl_c` is super nice, and I'll try to incorporate those in the current architecture!

> All `Session`s are stored in a `Arc<HashMap<String, Session>>`, with its name as the key. Since the hosts are loaded at the startup, it is perfectly OK to store them in a `Arc<HashMap>`.

I suppose it can even be a `Arc<Vec<Session>>` because sessions aren't really distinguished between each other.

> Then, upon receiving new tasks, search for the corresponding `Session` and spawn a new task to execute the job.

Similarly, there is no "corresponding" session. In broadcast most, all sessions will be used so that's not a problem, but in queue mode, we need to keep track of which sessions are currently free, and whenever a new command is fetched or a new session becomes free, we execute the command using the free session. To do so, we need to track the status (in use or free) of the sessions. From the top of my head, I suppose keeping `is_free: Arc<Mutex<Vec<bool>>>` and setting the flags at the beginning and end of each task would be an option, but that will require the scheduling loop to poll on the vector.

> The main task can then call [`tokio::sync::Semaphore::acquire_many`](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html#method.acquire_many) to wait for the completion of `n` tasks and uses [`tokio::sync::SemaphorePermit::forget`](https://docs.rs/tokio/latest/tokio/sync/struct.SemaphorePermit.html#method.forget) to prevent the permit from being returned to the [`tokio::sync::Semaphore`](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html).

Definitely works for broadcast mode but we need something more fine-grained than `acquire_many` for queue mode.

> By doing each job in a task instead of running every `Session` in a task, we can do graceful shutdown easily using `tokio::signal::ctrl_c`.

I suppose I should start by ditching `ctrlc` and use `tokio::signal::ctrl_c` right now:
```rust
// Replacing ctrlc::set_handler
tokio::spawn(async move {
    tokio::signal::ctrl_c().await;
    cancel_tx.send(()).await.unwarp();
});
```

> Using the architecture I described above, I believe it is also possible to use [`notify`](https://docs.rs/notify/4.0.17/notify/) in `tokio::spawn` combined with [`tokio::sync::watch`](https://docs.rs/tokio/latest/tokio/sync/watch/index.html) to monitor the change for change of `queue.yaml`.
> 
> It is possible since [`tokio::sync::watch::Receiver::changed`](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.changed) is cancel safe.

This is great. Now, instead of just waiting three seconds when `queue.yaml` is empty, I can await on `watcher_rx.changed()` and whenever it notifies me, `continue` to the next iteration which will immediately call `job_queue.next()`.

> Upon inspection of implementation of [`tokio::signal::ctrl_c`](https://docs.rs/tokio/latest/src/tokio/signal/unix.rs.html), I found that it just uses `tokio::sync::watch::Receiver` to receive signal, so it should also be cancel safe.

That is good to know, but I didn't quite follow why cancel safety for `ctrl_c` is needed in your proposed architecture.

> If we actually need to unload/load new hosts dynamically, then I would use `HashMap<String, Arc<Session>>` to stores the sessions.

This could be a pretty good potential feature. Although I currently don't have a need for this, I'll definitely agree if someone will add this feature.

Reply:
> Similarly, there is no "corresponding" session.

Oh right, I forgot this is the  broadcast mode where session doesn't matter.

> Definitely works for broadcast mode but we need something more fine-grained than acquire_many for queue mode.

We can have a `Vec<Arc<Session>>` for storing the all sessions and creates a [`tokio::sync::mpsc::channel`] for storing all `Session` that are ready.

I would initialize the `channel` like this:

```
let mut sessions = Vec::new();

session.push(Arc::new(...));
// ...

let (sender, mut receiver) = channel(sessions.len());

for session in sessions.iter() {
    sender.send(session.clone());
}
```

When creating a new task:

```
// Get a session that is free.
let session = receiver.recv().unwrap();

let sender = sender.clone();
tokio::spawn(async move {
    // do something...

    let res = session.run(...).await;
    // handle the error...
    
    sender.send(session).await;
});
```

Using this way, we now have fine-grained notification on which `Session` is free for queue mode.

> I suppose I should start by ditching ctrlc and use tokio::signal::ctrl_c right now:

Agreed.

> That is good to know, but I didn't quite follow why cancel safety for ctrl_c is needed in your proposed architecture.

I imagined we will be doing something like this:

```
tokio::select! {
    _ = ctrl_c() => // graceful shutdown,
   _ => wait_for_queue_to_change() => // do more work,
}
```

[`tokio::sync::mpsc::channel`]: https://docs.rs/tokio/latest/tokio/sync/mpsc/fn.channel.html

Reply:
So we're basically tossing around the `Session` object! I That's elegant. And we only have to use one channel. And we can still `into_iter` and iterate over the `sessions` vector to gracefully `close` all of them. I acknowledge that switching to your architecture will benefit Pegasus.

But still, as far as I can understand, the performance would not be significantly better (maybe spawning a new task for each command might slightly degrade performance?) and it doesn't enable something that we weren't able to accomplish with the current architecture. Since the architecture is quite simpler, I will definitely keep this in mind and consider refactoring to this when features are added and the code gets too dirty.

You know, I would have went through the refactoring process if it was Friday night, but... it's Monday soon.. haha.. 😢

Reply:
> But still, as far as I can understand, the performance would not be significantly better (maybe spawning a new task for each command might slightly degrade performance?) and it doesn't enable something that we weren't able to accomplish with the current architecture.

This is true, however I think a simpler architecture will always make things easier, regardless of whether it is to support new features, fix existing bugs or performance optimizations.

> You know, I would have went through the refactoring process if it was Friday night, but... it's Monday soon.. haha..

Totally understandable.

Reply:
> This is true, however I think a simpler architecture will always make things easier, regardless of whether it is to support new features, fix existing bugs or performance optimizations.

100% agree! Your proposal will not go void.